File: Regresion polinomica/polynomial_regression.R
# Regresion polinomica

#import dataset
dataset = read.csv("Position_Salaries.csv")
dataset = dataset[,2:3]

#divide dataset into training and testing
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split==TRUE)
# testing_set = subset(dataset, split==FALSE)

#Scaling values
#training_set[,2:3] = scale(training_set[,2:3])
#testing_set[,2:3] = scale(testing_set[,2:3]) 

#Ajustar el modelo de regresion lineal simple con el conjunto de datos
ln_reg=lm(formula = Salary ~ .,
             data = dataset)
summary(ln_reg)

#Ajustar el modelo de regresion polinomica con el conjunto de datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
dataset$Level5 = dataset$Level^5
poly_reg = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_reg)

#Visualizar la regresion lineal simple
library(ggplot2)
ggplot() +
  geom_point(aes(x=dataset$Level, y=dataset$Salary),
             colour="red") +
  geom_line(aes(x=dataset$Level,
                y=predict(ln_reg, newdata = dataset)),
            colour="blue") +
  ggtitle("Modelo de regresión lineal del sueldo en función del nivel del empleado")+
  xlab("Posición del empleado")+
  ylab("Sueldo ($)")

#Visualizar la regresion lineal polinomica
x_grid = seq(min(dataset$Level), max(dataset$Level),0.1) # para suavizar la curva
library(ggplot2)
ggplot() +
  geom_point(aes(x=dataset$Level, y=dataset$Salary),
             colour="red") +
  geom_line(aes(x=x_grid,
            y=predict(poly_reg, newdata = data.frame(Level = x_grid,
                                                     Level2 = x_grid^2, 
                                                     Level3 = x_grid^3, 
                                                     Level4 = x_grid^4, 
                                                     Level5 = x_grid^5))),
            colour="blue") +
  ggtitle("Modelo de regresión polinómico del sueldo en función del nivel del empleado")+
  xlab("Posición del empleado")+
  ylab("Sueldo ($)")

#Predecir resultados con el modelo de regresion lineal
y_pred = predict(ln_reg, newdata = data.frame(Level = 6.5))

#Predecir resultados con el modelo de regresion polinomica
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5, 
                                                     Level2 = 6.5^2, 
                                                     Level3 = 6.5^3, 
                                                     Level4 = 6.5^4, 
                                                     Level5 = 6.5^5))

File: Regresion polinomica/polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 22 13:22:11 2021

@author: axvargas
"""

# Regresion polinomica

#lybraries import
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
#data import
dataset = pd.read_csv('Position_Salaries.csv')

#Matrix de caracteristicas
X = dataset.iloc[:,1:2].values # Debido a que tiene que ser una matriz y no un vector OJO [:,1] vector. [,1:2] matriz: Pero contienen los mismos datos
y = dataset.iloc[:,-1].values

#No divido el dataset por que hay muy pocos datos

#Ajustar la regresion lineal con el dataset OJO solo para comparar
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)


#Ajustar la regresion polinomica con el dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 5) #Si se quiere probar un grado mas solo se le cambia
X_poly = poly_reg.fit_transform(X)

lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly,y)

#Visualizacion de los resultados Modelo lineal
plt.scatter(X, y, color="red")
plt.plot(X, lin_reg.predict(X), color="blue")
plt.title("Modelo de regresión lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo ($)")
plt.show()
#Visualizacion de los resultados Modelo polinomico
#Para suavizar la curva:
X_grid = np.arange(min(X), max(X)+0.1, 0.1)
X_grid  = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color="red")
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color="blue") #OJO que se usa X_poly para realizar la preiccion na mas
plt.title("Modelo de regresión polinómica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo ($)")
plt.show()

#PRedicciones de nuestro modelos puesto para region manager-Partner: 6-7
lin_reg.predict([[6.5]])
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))
